### PR TITLE
[AOTInductor] Double buffering for Weights

### PR DESCRIPTION
**Pull request by muchulee8**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114446

Summary:
This adds function to model container doing weight swapping with double buffering.

There are 2 parts for double buffering
a) Write constants into inactive buffer
b) Swap active buffer

For (a), we write the constants into the buffer that's currently not in use, and store the information in both constants map and the corresponding constant array to read.
For (b), we obtain the lock, and activate the constant map/constant array that is inactive, and flag the one that's currently in use to inactive.

Test Plan:
test/cpp/aot_inductor/test.cpp

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @aakhundov @ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

### CODE REVIEW

**Comment by muchulee8** [comment]
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51543732).

File: torch/csrc/inductor/aoti_runtime/model.h [hunk 1]
```diff
@@ -379,22 +390,43 @@ class AOTInductorModelBase {
...
     if (!constants_map_) {
-      return;
+      throw std::runtime_error{
+          "constants_map_ was not ready when constants_ is trying to be constructed from it!"};
+    }
+    if (!constants_) {
+      constants_ = std::make_shared<std::vector<AtenTensorHandle>>(
+          constants_info_.size());
+    } else {
+      constants_->resize(constants_info_.size());
```
[the remark this answers is not in the document]
Comment (muchulee8) [reply torch/csrc/inductor/aoti_runtime/model.h:402]: Yes. They are unchanged. So this is basically a no-op after constants get initiated.

This is actually just copying the original code (previously, `update_constants_map` does (1) update the constants map (2) Populate the corresponding tensor in constant map to the array.) Now I have moved (2) to this separate function (`update_constants_array_from_map`.)

File: torch/csrc/inductor/aoti_runtime/model_container.h [hunk 2]
```diff
@@ -89,6 +95,101 @@ class AOTInductorModelContainer {
...
+
+      // Move the data to container handled blob.
+      uint8_t* internal_constants_ptr =
+          constants_blob_ptr + constants_internal_offset_[idx];
+      void* user_constant_ptr;
+      int64_t constant_size;
+      aoti_torch_get_data_ptr(it->second, &user_constant_ptr);
+      aoti_torch_get_storage_size(it->second, &constant_size);
+
+      AOTI_RUNTIME_DEVICE_CHECK(cudaMemcpy(
```
[the remark this answers is not in the document]
Comment (muchulee8) [reply torch/csrc/inductor/aoti_runtime/model_container.h:130]: The latter question, whether it's on device or on host are both possible, and there's no trivial way to identify whether it's on device or on host.

For the former question, I did a copy for safeness. I don't think we can "steal the pointer", the underlying storage for the tensor is a raw pointer, unlike smart pointers, stealing the ownership afaik is not possible. We can definitely just "use the pointer", I don't see issue for this. But for such case, we also don't manage the data, and something might break if the data managed by the user get freed.

File: torch/csrc/inductor/aoti_runtime/model.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (chenyang78) [reply torch/csrc/inductor/aoti_runtime/model.h:402]: OK, I see. So, the else-branch is executed when we first initialize the ```constants``` vector.

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

**Review by eellison — COMMENTED** [review]
would you please update the summary with details of the pr ?

File: torch/csrc/inductor/aoti_runtime/model_container.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/inductor/aoti_runtime/model_container.h:130]: > The latter question, whether it's on device or on host are both possible, and there's no trivial way to identify whether it's on device or on host.

I feel like this should be possible...

> don't think we can "steal the pointer", the underlying storage for the tensor is a raw pointer, unlike smart pointer

If the input is an PyTorch tensor, the way you do this is by setting it's deallocation function function to be a no-op, and then manage our own deallocation function which would call into the caching allocation and free it's pointer.  See - code changing dealloc function [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/cuda/Module.cpp#L1099-L1107). 

Although, there might be some abi complications here, and we could do as a follow up.

File: torch/csrc/inductor/aoti_runtime/model_container.h [hunk 3]
```diff
@@ -89,6 +95,101 @@ class AOTInductorModelContainer {
...
+    auto num_constants = models_[0]->num_constants();
+    for (size_t idx = 0; idx < num_constants; idx++) {
+      constants_array->at(idx) =
+          constants_map->find(models_[0]->constant_name(idx))->second;
+    }
+  }
+
+  void swap_constant_buffer() {
+    std::lock_guard unique_lk(model_exec_mutex_);
+    // Need to wait?
```
Comment (eellison) [inline torch/csrc/inductor/aoti_runtime/model_container.h:180]: What do you mean by this `// Need to wait?`

File: test/cpp/aot_inductor/test.py [hunk 4]
```diff
@@ -30,10 +30,13 @@ def forward(self, x, y):
         ]
         model_so_path = aot_compile(model, (x, y), constraints=constraints)
 
+    params = dict(model.named_parameters())
     data.update({
         f"model_so_path_{device}": model_so_path,
         f"inputs_{device}": [x, y],
         f"outputs_{device}": [ref_output],
+        f"fc_weight_{device}": params["fc.weight"],
+        f"fc_bias_{device}": params["fc.bias"],
```
Comment (eellison) [inline test/cpp/aot_inductor/test.py:39]: Why was this necessary ?

File: torch/csrc/inductor/aoti_runtime/model_container.h [hunk 5]
```diff
@@ -89,6 +95,101 @@ class AOTInductorModelContainer {
...
+          cudaMemcpyDefault));
+
+      // Generate Tensor from container handled blob.
+      // We extract stride and offset from provided Tensor since we do not
+      // guarantee that the tensor is contiguous.
+      AtenTensorHandle tensor_handle;
+      int64_t* stride;
+      int64_t offset;
+      int device_idx = -1;
+      AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_get_strides(it->second, &stride));
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/inductor/aoti_runtime/model_container.h:143]: Maybe you're already doing this but add asserts offsets/strides are the same ?

**Review by eellison — APPROVED** [review]
Looks good

File: torch/csrc/inductor/aoti_runtime/model_container.h [hunk 6]
```diff
@@ -126,12 +227,25 @@ class AOTInductorModelContainer {
...
   CUDAPtr constant_blob_;
+  CUDAPtr constant_blob_secondary_;
+
+  // Let's place this within USE_CUDA at the moment before we fully support
+  // update for CPU cases.
+  size_t blob_size_;
+  std::vector<size_t> constants_internal_offset_;
 #endif // USE_CUDA
 
+  bool use_secondary_;
```
[the remark this answers is not in the document]
Comment (muchulee8) [reply torch/csrc/inductor/aoti_runtime/model_container.h:238]: Added.
